Editor: Contact Form: Add hover text to Edit button

nitrajka opened the pull request:
Hi,

I fixed the bug #4116 . I came up with 4 solutions, but decided to use the following one. 
I wrapped the `<Gridicon>` component in a new `<div>` and styled the `<div>` with paddings. Exactly like the remove button located in 'components/tinymce/plugins/contact-form/dialog/field-remove-button.jsx' which is on the left side of edit button. 
If there wouldn't be paddings, the hover text would appear not in the same line as the hover text for remove button. I added a Popover component to the edit button. If there would be problem with ES6 import which I added, I can rewrite it to es5 or the others rewrite to es6.

gwwar commented:
Hi @nitrajka thanks for taking a stab at this! This issue is actually a bit harder than it appears.

`client/components/foldable-card/index.jsx` is a generic component and is used in a lot of places so we'll need to take a slightly different approach in this case. If we used the current diff, this means we'll see that edit button in a lot of other places that we didn't expect.

A few approaches:
- Create another component similar to https://github.com/Automattic/wp-calypso/blob/master/client/components/tinymce/plugins/contact-form/dialog/field-remove-button.jsx in the same folder
- After we have this new component, we can pass through both buttons in field.jsx:
```javascript
			<FoldableCard
				summary={ bothButtons }
				expandedSummary={ bothButtons }
```

cc @retrofox, @rodrigoi and @aduth for second opinions.

nitrajka commented:
Hi @gwwar, 
I understood. I'm going to take a look at it. Thanks for review.

nitrajka commented:
While reading the documentation (README.md) of `components/foldable-card` I noticed that there is a possibility to pass the `edit-button-component` as a prop of `FoldableCard`. Notice the `actionButton` and the `actionButtonExpanded` props of `FoldableCard`. 
So I created new component `field-edit-button` according to the `field-remove-button` in the same file, styled it and passed it through as a prop of `FoldableCard`. I also removed the first commit changes. 
I'm sure that there will be a problem with styles of this new component, I'll fix it soon.

@gwwar

gwwar commented:
Sounds good. Ping me again when you're ready for another 👀 at the PR

nitrajka commented:
I refactored the styles of my new component - `FieldEditButton`. 
This my`FieldEditButton`, or whichever `Gridicon` is placed instead, is being rotated (transformed) every time on `expand` action. Watch example here: https://wpcalypso.wordpress.com/devdocs/design/foldable-card. 
And I think, we don't want the `pencil-gridicon` to rotate after expand.

So my solution is: 
If the `FoldableCard` is  expanded, I pass through its `actionButtonExpanded` prop `<FieldEditButton expanded={ true } />`. Else I pass through its `actionButton` prop `<FieldEditButton expanded={ false } />`. According to this `expanded` prop of `FieldEditButton` the `classNames` package decides, whether to use `is-expanded` className or not.
I styled this `is-expanded` class in `styles.scss` file. If I didn't make the CSS selector so long, other styles would apply on my component.

Now, I would like to rewrite this component in ES6 and refactor classNames but I'm not sure about some things.
1. I read in the wp-calypso CSS documentation that 'The component fragment matches the folder name of the React component it's providing styles to.'
I thought that, if I create the className of `FieldEditButton` similarly to `FieldRemoveButton` it will be OK. But now I don't think so. I think, the classNames which I'm using are too long, and not appropriate. Sorry, I have no idea, how should the classNames look like in this case. Could you help me, please?
2. To rewrite it, I need to implement PureRenderMixin somehow. Is it ok to use `react-addons-pure-render-mixin` package like there https://facebook.github.io/react/docs/pure-render-mixin.html ? Or I'll have to use PureComponent from this library https://github.com/gaearon/react-pure-render , which you are actually using?

Thanks for review @gwwar

gwwar commented:
> To rewrite it, I need to implement PureRenderMixin somehow.

React added PureComponent in 15.3.0, so we can import this via
```javascript
import import React, { PureComponent } from 'react';
```
see: https://github.com/Automattic/wp-calypso/pull/7599 for example usages

gwwar commented on client/components/tinymce/plugins/contact-form/style.scss:
```diff
@@ -78,3 +78,20 @@
+	&.is-expanded {
+		transform: rotate(0);
+	}
+}
```
Missing an end of line here 😄

gwwar commented on client/components/tinymce/plugins/contact-form/dialog/field-edit-button.jsx:
```diff
@@ -0,0 +1,55 @@
+import Gridicon from 'components/gridicon';
+import Popover from 'components/popover';
+
+export default React.createClass( {
```
We can write this as either a stateless functional component, or an es6 class that extends PureComponent.

```javascript
function ContactFormDialogFieldEditButton( { expanded } ) {
// returns the same jsx as what's in the current render() method
}
```

```javascript
export default class ContactFormDialogFieldEditButton extends PureComponent {
```

gwwar commented on client/components/tinymce/plugins/contact-form/dialog/field-edit-button.jsx:
```diff
@@ -0,0 +1,55 @@
+import Popover from 'components/popover';
+
+export default React.createClass( {
+	displayName: 'ContactFormDialogFieldEditButton',
```
We use a babel transform that picks this up from the class or function name. We generally don't need to set this unless it needs to be different from the class or function name.

gwwar commented on client/components/tinymce/plugins/contact-form/dialog/field-edit-button.jsx:
```diff
@@ -0,0 +1,55 @@
+
+	propTypes: {
+		expanded: PropTypes.bool.isRequired
+	},
```
For a stateless function we can add directly:
```javascript
ContactFormDialogFieldEditButton.propTypes = {

};
```

For a class this can be defined as a [static](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/static) property:
```javascript
static propTypes = {
    expanded: PropTypes.bool.isRequired
}
```

gwwar commented on client/components/tinymce/plugins/contact-form/dialog/field-edit-button.jsx:
```diff
@@ -0,0 +1,55 @@
+	},
+
+	render() {
+		const classes = classNames("editor-contact-form-modal-field__edit", {
```
We need another [space](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#arrays-and-function-calls) in:
```javascript
const classes = classNames( "editor-contact-form-modal-field__edit",
```

You can also see relevant style warnings locally by running
```
make eslint-branch 
```
Feel free to ignore existing warnings outside of the new `field-edit-button.jsx`

gwwar replied:
From the eslint warning, we're expecting a `dialog` prefix. cc @mtias should we ignore this one, or use something like `dialog-editor-contact-form-field__edit`?

gwwar commented on client/components/tinymce/plugins/contact-form/style.scss:
```diff
@@ -78,3 +78,20 @@
+.editor-contact-form-modal-field__edit-wrapper {
+	height: 100%;
+
+	button {
```
Do you mind adding another class to the button instead. We can then do something like:
```
.editor-contact-form-modal-field__edit-wrapper {
    height: 100%;
}
.editor-contact-form-modal-field__edit-wrapper-button {
   width: 100%;
}
```

gwwar commented on client/components/tinymce/plugins/contact-form/style.scss:
```diff
@@ -78,3 +78,20 @@
+		width: 100%;
+	}
+}
+.editor-contact-form-modal-field__edit-wrapper.foldable-card__expand .editor-contact-form-modal-field__edit {
```
We try to avoid this much specificity, but it might not be avoidable in this case.

gwwar commented on client/components/tinymce/plugins/contact-form/dialog/field.jsx:
```diff
@@ -72,15 +73,17 @@ export default React.createClass( {
 				expanded={ this.props.isExpanded }
 				onOpen={ () => this.props.onUpdate( { isExpanded: true } ) }
-				onClose={ () => this.props.onUpdate( { isExpanded: false } ) }>
+				onClose={ () => this.props.onUpdate( { isExpanded: false } ) }
```
Let's try to avoid the function creation during each render, we can do something like:
```
handleCardOpen() {
    this.props.onUpdate( { isExpanded: true } );
},
handleCardClose() {
    this.props.onUpdate( { isExpanded: false } );
}
//...
render() {
    //...
    <FoldableCard 
            onClose={ this.handleCardClose }
             onOpen={ this.handleCardOpen }
}
```

gwwar commented:
Thanks @nitrajka I tested this and it seems to work well. I left a few notes in case you feel up to rewriting this as a stateless function, or an es6 class.

cc @alisterscott for some help testing

nitrajka commented:
Thank you very much, @gwwar , for your time and advisements. :)

I'm almost finished with code refactoring. I'm rewriting it as ES6 class, but one thing is still bugged. 
The console informs me that `{ this.translate( 'Edit Field', { context: 'button tooltip' } ) }` function in the render method of my component is not defined. It means, the `FieldEditButton` component 'thinks', that the translate method is his own method because I use `this` keyword. How to fix it?

gwwar commented:
For any translate usages, we can use localize as a [higher order component]https://github.com/Automattic/i18n-calypso#localize] rather than a mixin.  The new component wraps the original component, passing all original props plus props to assist in localization.

```
import { localize } from 'i18n-calypso';
class ContactFormDialogFieldEditButton extends PureComponent {
    render() {
      const { translate } = this.props;
      return ( <div> { translate( 'Hello World!' ); } </div> ); 
    }
}
export default localize( ConstactFormDialogFieldEditButton );
```

nitrajka commented:
Thanks a lot, @gwwar!  :)
As for me, I'm done with the code, ready to merge. Test it.

gwwar commented on client/components/tinymce/plugins/contact-form/dialog/field-edit-button.jsx:
```diff
@@ -0,0 +1,54 @@
+				<Popover
+					isVisible={ this.state.showTooltip }
+					context={ this.refs && this.refs.editField }
+					onClose={ () => {} }
```
Pretty minor, but we can either define this outside of render, or use lodash noop
```
import { noop } from 'lodash';
//...
onClose= { noop }
```

gwwar commented on client/components/tinymce/plugins/contact-form/dialog/field-edit-button.jsx:
```diff
@@ -0,0 +1,54 @@
+	}
+
+	render() {
+		const { translate } = this.props;
```
Optional but, since we're destructuring we can also do:
```javascript
const { expanded, translate } = this.props;
//...
'is-expanded': expanded
```

gwwar commented on client/components/tinymce/plugins/contact-form/dialog/field-edit-button.jsx:
```diff
@@ -0,0 +1,54 @@
+					ref="editField"
+					borderless
+					onMouseEnter={ () => this.setState( { showTooltip: true } ) }
+					onMouseLeave={ () => this.setState( { showTooltip: false } ) }>
```
We can similarly write handlers once in the class rather than in render. The tricky part is that we don't have autobinding so we need to set that up in the constructor.
```
constructor() {
    super( ...arguments );
    this.state = { showTooltip: false };
    this.handleMouseEnter = this.handleMouseEnter.bind( this );
    this.handleMouseLeave = this.handleMouseLeave.bind( this );
}

handleMouseEnter() {
   this.setState( { showTooltip: true } );
}

handleMouseLeave() {
   this.setState( { showTooltip: false } );
}
```

gwwar commented:
This looks great ✨ @nitrajka! 😄 Almost there. I left a few more minor notes. 

If you could tidy up these last few comments, and squash a few of your commits, we'll be ready to :shipit:

alisterscott commented:
Thanks for working on this change. I tested this out and it worked well in Chrome, Firefox and Safari (OSX) - the only issue I found was in Safari, whilst the hover text looks great, the Edit button styling no longer displays the same as previously (and the other button)

![safari edit hover](https://cloud.githubusercontent.com/assets/128826/18457097/58a3ae6e-7999-11e6-9281-f77b86bce88b.png)

nitrajka commented:
Finally,
I refactored the code according to the @gwwar's suggestions and fixed the layout of `FieldEditButton` so it appears to look the same in Safari, Firefox and Chrome.

@alisterscott

gwwar commented:
Hi @nitrajka! The code looks good here 👍 Do you mind [squashing](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) down the code with a single [meaningful commit message](https://robots.thoughtbot.com/5-useful-tips-for-a-better-commit-message). I'll be able to merge this for you.

@mtias did you have any unaddressed concerns?

nitrajka commented:
Sorry, I didn't realize that you wanted to squash ALL of the commits into a SINGLE commit. I did it. Thank you for the link to the article about commit message. I hope my commit message is meaningful enough. :) @gwwar

gwwar commented:
Awesome, thanks @nitrajka for the contribution!
